public/app/features/explore/hooks/useExplorePageTitle.ts

```diff
@@ -1,16 +1,59 @@
-import { isTruthy } from '@grafana/data';
+import { useEffect, useRef } from 'react';
+
+import { NavModel } from '@grafana/data';
 import { Branding } from 'app/core/components/Branding/Branding';
 import { useNavModel } from 'app/core/hooks/useNavModel';
-import { useSelector } from 'app/types';
+import { safeParseJson } from 'app/core/utils/explore';
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { ExploreQueryParams } from 'app/types';
+
+export function useExplorePageTitle(params: ExploreQueryParams) {
+  const navModel = useRef<NavModel>();
+  navModel.current = useNavModel('explore');
+  const dsService = useRef(getDatasourceSrv());
+
+  useEffect(() => {
+    if (!params.panes || typeof params.panes !== 'string') {
+      return;
+    }
+
+    Promise.allSettled(
+      Object.values(safeParseJson(params.panes)).map((pane) => {
+        if (
+          !pane ||
+          typeof pane !== 'object' ||
+          !hasKey('datasource', pane) ||
+          !pane.datasource ||
+          typeof pane.datasource !== 'string'
+        ) {
+          return Promise.reject();
+        }
 
-import { selectOrderedExplorePanes } from '../state/selectors';
+        return dsService.current.get(pane.datasource);
+      })
+    )
+      .then((results) => results.filter(isFulfilled).map((result) => result.value))
+      .then((datasources) => {
+        if (!navModel.current) {
+          return;
+        }
 
-export function useExplorePageTitle() {
-  const navModel = useNavModel('explore');
+        const names = datasources.map((ds) => ds.name);
 
-  const datasourceNames = useSelector((state) =>
-    Object.values(selectOrderedExplorePanes(state)).map((pane) => pane?.datasourceInstance?.name)
-  ).filter(isTruthy);
+        if (names.length === 0) {
+          global.document.title = `${navModel.current.main.text} - ${Branding.AppTitle}`;
+          return;
+        }
 
-  document.title = `${navModel.main.text} - ${datasourceNames.join(' | ')} - ${Branding.AppTitle}`;
+        global.document.title = `${navModel.current.main.text} - ${names.join(' | ')} - ${Branding.AppTitle}`;
+      });
+  }, [params.panes]);
 }
+
+// TS<5 does not support `in` operator for type narrowing. once we upgrade to TS5, we can remove this function and just use the in operator instead.
+function hasKey<K extends string, T extends object>(k: K, o: T): o is T & Record<K, unknown> {
+  return k in o;
+}
+
+const isFulfilled = <T>(promise: PromiseSettledResult<T>): promise is PromiseFulfilledResult<T> =>
```

Comment:
`isFullfilled` already exists in `useStateSync`. Could we reuse it?

Reply:
done in 2e5f41a155